### PR TITLE
Upgrade Ava and @sindresorhus/tsconfig 

### PR DESCRIPTION
**Pull request by ClementValot**
Following discussions on https://github.com/sindresorhus/got/issues/2051

- Upgrade `ava` to V4 
- Minor packages upgrades
- Removing tsconfig property noPropertyAccessFromIndexSignature
- Upgrade `@sindresorhus/tsconfig` to 3.0.1 

#### Checklist

- [ X ] I have read the documentation.
- [ X ] I have included a pull request description of my changes.

### CODE REVIEW

File: source/core/index.ts [hunk 1]
```diff
@@ -103,15 +103,15 @@ export type GotEventFunction<T> =
...
+	 */
 	& ((name: 'uploadProgress' | 'downloadProgress', listener: (progress: Progress) => void) => T)
 	/**
 	To enable retrying on a Got stream, it is required to have a `retry` handler attached.
 
 	When this event is emitted, you should reset the stream you were writing to and prepare the body again.
 
 	See `got.options.retry` for more information.
-	*/
+	 */
```
Comment (sindresorhus) [inline source/core/index.ts:114]: Don't make unrelated changes

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: tsconfig.json [hunk 2]
```diff
@@ -2,12 +2,10 @@
 	"extends": "@sindresorhus/tsconfig",
 	"compilerOptions": {
 		"outDir": "dist",
-		"target": "es2020", // Node.js 14
-		"lib": [
-			"es2020"
+	    "moduleResolution": "Node",
```
Comment (sindresorhus) [inline tsconfig.json:5]: This should not be here

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: source/core/index.ts (hunk 1 above)
Comment (ClementValot) [reply source/core/index.ts]: IDE auto-indent shenanigans :( 
Fixed

File: tsconfig.json (hunk 2 above)
Comment (ClementValot) [reply tsconfig.json:6]: Atm got 12.3.1 has `moduleResolution=Node` inherited from `@sindresorhus/tsconfig 2.0.0`

Switching to `@sindresorhus/tsconfig 3.0.1` sets `moduleResolution` to `Node16` which breaks type checking when building got (due to issues with how some packages interact with `moduleReolution=Node16`, see debugging attempts in https://github.com/sindresorhus/got/issues/2051#issuecomment-1214196483 )

Overriding moduleResolution to Node here allows to upgrade tsconfig without breaking the package. 
It should ultimately be removed when the resolution with the problematic packages is okay. 

I added a comment in the file, does that sound reasonable?

**Comment by ClementValot** [comment]
I botched unit tests because ava cannot interpret typescript transpiled with module:Node16 
I fixed it by adding module:ES2020 to ts-node. 

Quick question @szmarczak, I see you're a maintainer of https://github.com/avajs/typescript, should it be added to this project? 

Also unit test can't run properly on my setup, but there are a few that actually broke due to a FakeTimer issue, looking into it  
```
  timeout › lookup timeout

  Rejected promise returned by test. Reason:

  TypeError {
    message: 'Can\'t install fake timers twice on the same global object.',
  }

  › Object.install (node_modules/@sinonjs/fake-timers/src/fake-timers-src.js:1642:19)
  › exec (file:///test/helpers/with-server.ts:18:38)

```

**Comment by sindresorhus** [comment]
> Minor packages upgrades
Removing tsconfig property noPropertyAccessFromIndexSignature

These are unrelated changes and should be done in a separate pull request.

**Comment by sindresorhus** [comment]
> Quick question @szmarczak, I see you're a maintainer of https://github.com/avajs/typescript, should it be added to this project?

`ts-node` is better in our case as it means it will not need to rewrite paths and stuff.

**Comment by sindresorhus** [comment]
Closing in favor of https://github.com/sindresorhus/got/commit/86308155d1bc43c635773183db908af902922549 and https://github.com/sindresorhus/got/issues/2051#issuecomment-1236513684. Help welcome getting the dependencies fixed.
